etl/steps/__init__.py

```diff
@@ -98,6 +88,52 @@ def to_dependency_order(
     return filtered
 
 
+def filter_to_subgraph(
+    graph: Graph, includes: Iterable[str], incl_forward: bool = True
+) -> Graph:
+    """
+    Filter the full graph to only the included nodes, and all their dependencies.
+
+    If the incl_forward flag is true, also include "forward dependencies" (dependents),
+    and their own (backward) dependencies.
+
+    Assumes that the graph is organized dependent -> dependency (A -> B means A is
+    dependent on B).
+    """
+    all_steps = graph_nodes(graph)
+    subgraph: Graph = defaultdict(set)
+
+    included = {
+        s for s in all_steps if any(re.findall(pattern, s) for pattern in includes)
+    }
+
+    if incl_forward:
+        # Reverse the graph to find all nodes dependent on included nodes (forward deps)
+        forward_deps = set(traverse(reverse_graph(graph), included))
+        included = included.union(forward_deps)
+
+    # Now traverse the other way to find all dependencies of included nodes (backward deps)
+    return traverse(graph, included)
```

Comment:
Let me know if I missed something, but I'm pretty sure from my unit tests that the 3 separate BFS's + parent_frontier/child_frontier stuff isn't necessary. We can get the same result by simply BFS'ing in each direction, and we can also make the BFS generic, so I did that here

Reply:
(In addition to the unit test, I also tested that `etl data://garden/wb/2021-07-01/wb_income --dry-run` seems to generate the same 28 steps before and after my change)